libs/metrics-core/src/Data/Metrics.hs

```diff
@@ -2,214 +2,281 @@
 gaugeSet x p m = liftIO $ do
-    Gauge g <- gaugeGet p m
-    -- To play it safe, we want a full memory barrier, which 'writeCounter'
-    -- does not provide, so we use a CAS loop instead. This is not worse
-    -- than e.g. a CAS loop inherent to a comparable 'atomicModifyIORef'.
-    casLoop g =<< Atomic.readCounterForCAS g
-  where
-    casLoop g v = do
-        (ok, v') <- Atomic.casCounter g v x
-        unless ok (casLoop g v')
+    g <- gaugeGet p m
+    P.setGauge g x
 
-gaugeAdd :: MonadIO m => Int -> Path -> Metrics -> m ()
+-- | Add the given amount to the gauge at 'Path'
+gaugeAdd :: MonadIO m => Double -> Path -> Metrics -> m ()
 gaugeAdd x p m = liftIO $ do
-    Gauge g <- gaugeGet p m
-    Atomic.incrCounter_ x g
+    g <- gaugeGet p m
+    P.addGauge g x
 
+-- | Add 1 to the gauge at 'Path'
 gaugeIncr :: MonadIO m => Path -> Metrics -> m ()
 gaugeIncr = gaugeAdd 1
 
+-- | Subtract 1 to the gauge at 'Path'
 gaugeDecr :: MonadIO m => Path -> Metrics -> m ()
 gaugeDecr = gaugeAdd (-1)
 
-gaugeSub :: MonadIO m => Int -> Path -> Metrics -> m ()
+-- | Subtract the given amount from the gauge at 'Path'
+gaugeSub :: MonadIO m => Double -> Path -> Metrics -> m ()
 gaugeSub x = gaugeAdd (-x)
 
-gaugeValue :: MonadIO m => Gauge -> m Int
-gaugeValue (Gauge g) = liftIO $ Atomic.readCounter g
+-- | Get the current value of the Gauge
+gaugeValue :: MonadIO m => Gauge -> m Double
+gaugeValue g = liftIO $ P.getGauge g
 
 -----------------------------------------------------------------------------
--- Label specifics
-
-labelGet :: MonadIO m => Path -> Metrics -> m Label
-labelGet t m = liftIO $ do
-    ls <- readIORef (labels m)
-    maybe make return (Map.lookup t ls)
-  where
-    make = do
-        l <- Label <$> newIORef T.empty
-        atomicModifyIORef' (labels m) $ \ls -> (Map.insert t l ls, l)
-
-labelGet' :: MonadIO m => Path -> Metrics -> m Label
-labelGet' t m = liftIO $ do
-    l <- Label <$> newIORef T.empty
-    atomicModifyIORef' (labels m) $ \ls ->
-        case Map.lookup t ls of
-            Nothing -> (Map.insert t l ls, l)
-            Just l' -> (ls , l')
-
-labelSet :: MonadIO m => Text -> Path -> Metrics -> m ()
-labelSet x p m = liftIO $ do
-    Label l <- labelGet p m
-    atomicModifyIORef' l $ const (x, ())
-
-labelValue :: MonadIO m => Label -> m Text
-labelValue (Label l) = liftIO $ readIORef l
-
------------------------------------------------------------------------------
--- Buckets specifics
-
-bucketsGet :: MonadIO m => Int -> Int -> Path -> Metrics -> m Buckets
-bucketsGet k n t m = liftIO $ do
-    bs <- readIORef (buckets m)
-    maybe make return (Map.lookup t bs)
+-- Histogram specifics
+
+-- | *DEPRECATED*
+-- These are the exact histogram bucket markers which the old *custom* metrics-core
+-- library used. Some grafana graphs are still built around these exact number
+-- e.g. see gally's POST duration graph:
+--   https://staging-ie-grafana.zinfra.io/dashboard/db/galley
```

Comment:
This link points into our internal infrastructure.  I don't mind it being here since it's just legacy code, but should be marked as "ignore if you can't reach this" somehow.

Reply:
Ahh; good point; I'll put it in a code comment but not in haddocks maybe?